### DISCUSSION TITLE
[PrimeReact] Styles problems with NextJS + Tailwind

### DISCUSSION THREAD

**Discussion by SoyDiego** [primefaces/community#294] [PrimeReact] [answered]
Hi, It's my first time using PrimeReact with NextJS and Tailwind and I have problem in my styles and I don't understand why.
Here you can see a **Stackblitz** with the problem: https://stackblitz.com/edit/stackblitz-starters-7lccbl?file=app%2Fpage.tsx

### I realized if you "play" with the styles and hide some of them you can "repair" a little bit but broke others. I think the styles are being stepped

![problem primereact](https://github.com/primefaces/community/assets/19764334/1a95b7bb-6c45-44b9-9e90-d7daa49b8e6f)

## Problem 1 (Opening **Stackblitz**)

If I open in Chrome (left side) works perfectly. But If I open in Firefox (right side) it's broken. (**Button** "Volver atrás", **calendar input**, **dividers**, and **button at the bottom)**

![image](https://github.com/primefaces/community/assets/19764334/e8694d91-dc4e-43bb-b68e-00c4b2ac46d4)

## Problem 2 (Opening **Netlify**)

I uploaded almost the same code to Netlify to check it and in Chrome and Firefox it's broken.
This is the link: https://lighthearted-genie-243620.netlify.app/pages/salvaments.

## Problem 3 (Opening **VERCEL**)
Here you can see the problem too: https://estadisticas-guardavidas.vercel.app/pages/salvaments

![image](https://github.com/primefaces/community/assets/19764334/43dde848-8c13-4258-b652-76a9be2fa5a2)

Hope someone knows If i'm doing something wrong or it's a bug.

Thanks!

**Comment by melloware** [comment]
I don't know why but it looks like Firefox and Chrome are interpreting the `layers` differently or ordering of `@layer primreact` and `@layer tailwind`

**Reply by melloware** [reply, marked as answer]
OK all examples updated to Next 15 and Tailwind4: https://github.com/primefaces/primereact-examples

**Comment by SoyDiego** [comment]
> I don't know why but it looks like Firefox and Chrome are interpreting the `layers` differently or ordering of `@layer primreact` and `@layer tailwind`

Yep, but if you check the Netlify example, in both cases it's broken too, Chrome, Firefox.

**Reply by melloware** [reply]
OK so i wouldn't trust StackBlitz then sometimes it can be funky but a real running app is the real test.  My guess is its some kind of ordering thing when your app gets bundled its like its bundling the CSS in the wrong order?

**Comment by SoyDiego** [comment]
> OK so i wouldn't trust StackBlitz then sometimes it can be funky but a real running app is the real test. My guess is its some kind of ordering thing when your app gets bundled its like its bundling the CSS in the wrong order?

Maybe yes but I tried moving the order of the CSS layers and always the same result.
Here you have another example with VERCEL: https://estadisticas-guardavidas.vercel.app/pages/salvaments (I added to the first message)

**Reply by melloware** [reply]
Can you add all this detail and your samples to this ticket: https://github.com/primefaces/primereact/issues/4987

**Reply by SoyDiego** [reply]
Done bro! https://github.com/primefaces/primereact/issues/4987#issuecomment-1777092798

**Reply by SoyDiego** [reply]
I updated the first post and I left a gif, maybe can be useful, I don't know @melloware

**Reply by SoyDiego** [reply]
I found this possible solution: https://soydiego.notion.site/PrimeReact-How-to-fix-styles-NextJS-Tailwind-92bec3eb91594388a7be67d922460eaf

**Reply by bloodykheeng** [reply]
This solution is not working in next 15

**Reply by melloware** [reply]
@bloodykheeng see this solution for Next 15 https://github.com/primefaces/primereact/issues/5291#issuecomment-2634882102

**Reply by bloodykheeng** [reply]
@melloware ive tried setting up primereact with tailwind in next 15 as the docs say but seems tailwind is taking over most of prime react styles

**Reply by melloware** [reply]
Yes did you read that comment a user posted how it's different for 15 and how to make it work.

**Reply by bloodykheeng** [reply]
ive not seen it anywhere

**Reply by bloodykheeng** [reply]
@melloware  I'm the one who made that comment but it was i was talking about theming not styling the issue am facing now is that tailwind is not working hand in hand with prime react styles

**Reply by melloware** [reply]
Look two posts up I just posted a link!!!

**Reply by bloodykheeng** [reply]
@melloware are u meaning this [link from diego](https://soydiego.notion.site/PrimeReact-How-to-fix-styles-NextJS-Tailwind-92bec3eb91594388a7be67d922460eaf)

**Reply by melloware** [reply]
Yes PrimeReact has not been updated for 15 yet

**Reply by bloodykheeng** [reply]
@melloware what ive seen primereact cant work hand in hand with tailwind in next 15 ive followed all solutions people are saying but still most prime react components are breaking or dont get there styles well they come half backed i get buttons with no colours inputs with no borders etc

**Reply by melloware** [reply]
its how NextJS packages....the bundler its loading the CSS in a different order is my guess.

**Reply by bloodykheeng** [reply]
@melloware  incase of any solution in next15 u alert me but for the meantime ive uninstalled tailwind in my project and resorted to prime flex css

**Reply by bloodykheeng** [reply]
@melloware any solution yet for primereact react to work hand in hand with tailwind css in nextjs 15 🙃

**Reply by melloware** [reply]
I just updated the examples for tailwind4 

https://github.com/primefaces/primereact-examples   

I need to try and update one for Next15 next

**Reply by bloodykheeng** [reply]
Ok cool thats great to here i will be testing it out in my next project 😁😁

**Reply by melloware** [reply]
@bloodykheeng ok all updated to Next15 and Tailwind4 in all examples!  See: https://github.com/primefaces/primereact-examples

**Reply by bloodykheeng** [reply]
@melloware thanks very much for the great work & your team now tailwind v4 works fine with primereact in next 15

**Comment by bloodykheeng** [comment]
@soyDiego this is how i solved styles issue in next15 , primereact and tailwind also theming there is and issue of flickering

for flickering i stopped using the next themes package i realized it uses local storage which is not good with ssr so i resorted to my own way to use cookies

so for theming to work install prime react and copy the css files this is how i archived it on windows PowerShell

```bash
// To copy only lara-light-blue and lara-dark-blue from node_modules
// into your public/themes folder using PowerShell, follow these steps:
// -----------------------------------------------------------------------

// # Create destination directory if it doesn't exist
// New-Item -Path "public/themes" -ItemType Directory -Force

// # Copy lara-light-blue
// Copy-Item -Path "node_modules/primereact/resources/themes/lara-light-blue" -Destination "public/themes" -Recurse -Force

// # Copy lara-dark-blue
// Copy-Item -Path "node_modules/primereact/resources/themes/lara-dark-blue" -Destination "public/themes" -Recurse -Force
```

then now import prime react in  globals.css 

globals.css
```css
@import "tailwindcss";

@source 'primereact/resources/themes/saga-blue/theme.css';

/* Prime icons*/
@import "primeicons/primeicons.css";

@custom-variant dark (&:where(.dark, .dark *));

.card {
  padding: 2em;
}

```

then now create a Theme Provider and a Prime react Provider
```js
"use client";

import { createContext, useState, useEffect, useContext } from "react";
import Cookies from "js-cookie";

type Theme = "light" | "dark";
type ThemeContextType = {
    theme: Theme;
    setTheme: (theme: Theme) => void;
};

const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export function ThemeProvider({ children, defaultTheme }: { children: React.ReactNode, defaultTheme: any }) {

    const savedTheme = Cookies.get("theme") as Theme;
    const [theme, setTheme] = useState<Theme>(defaultTheme ?? savedTheme ?? "light");

    useEffect(() => {
        const html = document.documentElement;
        if (savedTheme) {
            setTheme(savedTheme);
        } else {
            Cookies.set("theme", theme, { expires: 365 }); // set initial default theme
        }

        if (savedTheme === "dark") {
            html.classList.add("dark");
            html.classList.remove("light");
            html.style.colorScheme = "dark";
            html.setAttribute("data-theme", 'dark'); // Set data-theme attribute
        } else {
            html.classList.add("light");
            html.classList.remove("dark");
            html.style.colorScheme = "light";
            html.setAttribute("data-theme", 'light'); // Set data-theme attribute
        }
    }, []);

    const handleSetTheme = (newTheme: Theme) => {
        setTheme(newTheme);
        Cookies.set("theme", newTheme, { expires: 365 }); // Store for a year

        const html = document.documentElement;

        if (newTheme === "dark") {
            html.classList.add("dark");
            html.classList.remove("light");
            html.style.colorScheme = "dark";
            html.setAttribute("data-theme", 'dark'); // Set data-theme attribute
        } else {
            html.classList.add("light");
            html.classList.remove("dark");
            html.style.colorScheme = "light";
            html.setAttribute("data-theme", 'light'); // Set data-theme attribute
        }
    };

    return (
        <ThemeContext.Provider value={{ theme, setTheme: handleSetTheme }}>
            {children}
        </ThemeContext.Provider>
    );
}

export function useTheme() {
    const context = useContext(ThemeContext);
    if (!context) {
        throw new Error("useTheme must be used within a ThemeProvider");
    }
    return context;
}

```


prime react provider
```js
"use client";
import { useEffect, useContext } from "react";

import { PrimeReactProvider } from 'primereact/api';
import { PrimeReactContext } from "primereact/api";
import { useTheme } from "./ThemeProvider";

function ThemeSync() {
    const { theme } = useTheme(); // Get theme from next-themes

    console.log("🚀 ~ ThemeSync ~ theme:", theme)
    const { changeTheme } = useContext(PrimeReactContext);

    useEffect(() => {
        if (!theme || !changeTheme) return; // Avoid SSR issues

        const linkId = "theme-link";
        const newTheme = theme === "dark" ? "lara-dark-blue" : "lara-light-blue";
        const currentTheme = newTheme === "lara-dark-blue" ? "lara-light-blue" : "lara-dark-blue";

        changeTheme(currentTheme, newTheme, linkId, () => {
            const existingLinks = document.querySelectorAll(`link[id="${linkId}"]`);
            if (existingLinks.length > 1) {
                document.head.removeChild(existingLinks[0]); // Remove old theme link
            }
        });
    }, [theme, changeTheme]);

    return null;
}


export default function PrimeReactProviders({ children }: { children: React.ReactNode }) {


    return (
        <PrimeReactProvider value={{ unstyled: false, }}>
            <ThemeSync />
            {children}
        </PrimeReactProvider>
    );
}


// To copy only lara-light-blue and lara-dark-blue from node_modules
// into your public/themes folder using PowerShell, follow these steps:
// -----------------------------------------------------------------------

// # Create destination directory if it doesn't exist
// New-Item -Path "public/themes" -ItemType Directory -Force

// # Copy lara-light-blue
// Copy-Item -Path "node_modules/primereact/resources/themes/lara-light-blue" -Destination "public/themes" -Recurse -Force

// # Copy lara-dark-blue
// Copy-Item -Path "node_modules/primereact/resources/themes/lara-dark-blue" -Destination "public/themes" -Recurse -Force

```


then go ahead and use them in your layout.tsx

```js
import type { Metadata } from "next";
import "./globals.css";
import "primereact/resources/themes/lara-light-blue/theme.css";

import { ThemeProvider } from '@/providers/ThemeProvider'
import PrimeReactProvider from "@/providers/PrimeReactProvider"

import { cookies } from "next/headers";
import NextJsProgressBar from '@/utils/NextJsProgressBar'

import { Inter } from "next/font/google";

const inter = Inter({ subsets: ["latin"] });

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {

  const getCookie = async () => {
    const cookieStore = await cookies()
    return cookieStore.get("theme")
  }

  const defaultTheme = getCookie();

  return (
    <html lang="en">
      <head>
        <link id="theme-link" rel="stylesheet" href="/themes/lara-light-blue/theme.css" />
      </head>
      <body className={`bg-[#FCFCFC] dark:bg-black ${inter.className}`}>
        <NextJsProgressBar />
        <ThemeProvider defaultTheme={defaultTheme}>
          <PrimeReactProvider>
            {children}
          </PrimeReactProvider>
        </ThemeProvider>
      </body>
    </html>
  );
}

```

now u good to go 😄
